server/src/main/java/com/coincoinche/config/DataSourceConfig.java

```diff
@@ -0,0 +1,52 @@
+package com.coincoinche.config;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import org.apache.tomcat.jdbc.pool.DataSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Profile;
+
+/** Programmatic configuration of the database connection. */
+@Configuration
+public class DataSourceConfig {
+
+  private static final Logger logger = LoggerFactory.getLogger(DataSourceConfig.class);
+
+  /**
+   * Programmatically configure a connection to the Heroku Postgres database.
+   * This is only avaiable for the heroku application profile.
+   * @return the datasource used for the connection.
+   */
+  @Bean
+  @Profile("heroku")
+  public DataSource postgresHerokuDataSource() {
+    String databaseUrl = System.getenv("DATABASE_URL");
+    logger.info("Initializing PostgreSQL database: {}", databaseUrl);
+
+    URI dbUri;
+    try {
+      dbUri = new URI(databaseUrl);
+    } catch (URISyntaxException e) {
+      logger.error(String.format("Invalid DATABASE_URL: %s", databaseUrl), e);
+      return null;
+    }
+
+    String username = dbUri.getUserInfo().split(":")[0];
+    String password = dbUri.getUserInfo().split(":")[1];
+    String dbUrl = "jdbc:postgresql://" + dbUri.getHost() + ':' + dbUri.getPort() + dbUri.getPath();
```

Comment:
What is the difference between dbUrl and dbUri in the end?

Reply:
`dbUrl` is a string while `dbUri` is an `URI` object. Also, `dbUrl` is formatted with the `jdbc` format while `dbUri` (which corresponds to `databaseUrl`) is formatted with another format.

Reply:
basically, this method builds the jdbc-formatted URL from the URL provided by Heroku.